src/java/org/apache/cassandra/schema/PartitionDenylist.java

```diff
@@ -0,0 +1,406 @@
+    {
+        return loadAttempts;
+    }
+    public synchronized int getLoadSuccesses()
+    {
+        return loadSuccesses;
+    }
+
+    /**
+     * Performs initial load of the partition denylist.  Should be called at startup and only loads if the operation
+     * is expected to succeed.  If it is not possible to load at call time, a timer is set to retry.
+     */
+    public void initialLoad()
+    {
+        if (!DatabaseDescriptor.enablePartitionDenylist())
+            return;
+
+        synchronized (this)
+        {
+            loadAttempts++;
+        }
+
+        // Check if there are sufficient nodes to attempt reading all the denylist partitions before issuing the query.
+        // The pre-check prevents definite range-slice unavailables being marked and triggering an alert.  Nodes may still change
+        // state between the check and the query, but it should significantly reduce the alert volume.
+        String retryReason = "Insufficient nodes";
+        try
+        {
+            if (readAllHasSufficientNodes() && load())
+            {
+                return;
+            }
+        }
+        catch (Throwable tr)
+        {
+            logger.error("Failed to load partition denylist", tr);
+            retryReason = "Exception";
+        }
+
+        // This path will also be taken on other failures other than UnavailableException,
+        // but seems like a good idea to retry anyway.
+        int retryInSeconds = DatabaseDescriptor.getDenylistInitialLoadRetrySeconds();
+        logger.info(retryReason + " while loading partition denylist cache.  Scheduled retry in {} seconds.",
+                    retryInSeconds);
+        ScheduledExecutors.optionalTasks.schedule(this::initialLoad, retryInSeconds, TimeUnit.SECONDS);
+    }
+
+    private boolean readAllHasSufficientNodes()
+    {
+        return RangeCommands.sufficientLiveNodesForSelectStar(SystemDistributedKeyspace.PartitionDenylistTable,
+                                                              DatabaseDescriptor.denylistConsistencyLevel());
+    }
+
+    public boolean load()
+    {
+        final long start = System.currentTimeMillis();
+        final Map<TableId, DenylistEntry> allDenylists = readAll();
+        if (allDenylists == null)
+            return false;
+
+        // Cache iterators are weakly-consistent
+        for (final Iterator<TableId> it = denylist.asMap().keySet().iterator(); it.hasNext(); )
+        {
+            final TableId tid = it.next();
+            if (!allDenylists.containsKey(tid))
+                allDenylists.put(tid, new DenylistEntry());
+        }
+        denylist.putAll(allDenylists);
+        logger.info("Loaded partition denylist cache in {}ms", System.currentTimeMillis() - start);
+
+        synchronized (this)
+        {
+            loadSuccesses++;
+        }
+        return true;
+    }
+
+    public boolean denylist(final String keyspace, final String cf, final ByteBuffer key)
```

Comment:
Maybe I missed it, but I found no method to remove a key from the denylist. Is there any? I think it is important to have it.

Reply:
Just to highlight: when we mark a key as denied and then unmark, it may take sometime while it expires from denylist caches. I'm not sure it is a problem, but we definitely need to mention that in documentation.

Reply:
I believe the UX in the current form is "It's easier to deny list (JMX or CQL) than it is to un deny something (CQL only". Debateable as to whether that's a friendly interface or not; it in fact seems somewhat _unfriendly_ to me, as well as annoyingly asymmetric. I'll take a pass early next week at also adding a "remove deny listed entry" operation via JMX.

Also, seems like we'd want to proactively remove the denied partition from the cache rather than waiting for the full refresh time.

Reply:
> Also, seems like we'd want to proactively remove the denied partition from the cache rather than waiting for the full refresh time.

I feel you meant "**un**denied partition". I agree, but I'm not sure how to do that because this operation needs to be propagated to other nodes, I'm not aware of such a mechanism in Cassandra (maybe I just don't know). Please, share your thoughts on the approach.

Reply:
Sorry, temporally the partition which is denied, transitioning to an undenied state, will live on in the cache as denied until such time as the cache reloads. We should tombstone that record on the local node at least and remove it from the cache on the JMX path.

We _could_ wire this into the Keyspace.applyInternal path, checking on a mutation if it's for the system_distributed.partition_denylist table. That way any change to the underlying denylist table would be immediately reflected in the local cache as well. I'm immediately skeptical of this as it's 1) the most obvious answer having worked on this codebase for awhile, and 2) endangers us turning Keyspace.applyInternal into the dumping ground to keep things in sync between state in memory and cql tables, and it's very much in the hot path. I'll spend a little more time tomorrow chewing on this question.

Reply:
Nope; scratch that above idea. We partition on a composite of ks_name + cf_name and cluster on partition, so we wouldn't expect data in the underlying CQL table for a denylist entry to correspond to the node that actually owns the access to that data.

Reply:
> Use the JMX hook to loadPartitionDenylist across the cluster

I feel we can invalidate/reload a single key instead of reloading the whole cache. But in general the approach makes sense to me! :+1:

Reply:
> a single key instead of reloading the whole cache
Currently the granularity is per-table.

The DenylistEntry was originally designed as an immutable set of keys and tokens so we'd have to break down that data structure and/or make it mutable in order to allow per-partition key refreshing. My guess is that, given the sizes of data we expect to be working with here (default 1k per table, 10k total), breaking down to per-pk refresh vs. per-table would be a premature optimization. I vote we leave it as is and keep that as an exercise for the future if we find we need that. What do you think?